### DISCUSSION TITLE
Use "named id" instead of generic "!{class}{n}"

### DISCUSSION THREAD

**Discussion by BloodyRain2k** [alejandroautalan/pygubu-designer#212] [Ideas]
Maybe I'm overlooking something here, but it seems to me that the only use for giving a widget a custom id, is for asking the builder for that widget later.

When going through the children of widgets or such, it's still the same `!label2` or whatever the builder ends up assigning to it.

For example, here I got an app that contains a label and a scrolled frame in the mainwindow (tk.Toplevel), both have names and both names are nowhere to be found:
![image](https://github.com/alejandroautalan/pygubu-designer/assets/1014362/ddc41ccb-db12-4a67-9170-956b09b1c1e1)

The `nametowidget()` method would be probably easier to use, if I could tell it select `.!results.!entry3.!file` and it would go from the mainwindow to "results" (the scrolled frame), then the 4th "entry" (simple frame) and the label "file" in that.
Currently the actual query for that would be `.!tkscrolledframe.!frame3.!label`. And yeah that'd work the same, but looking at widgets and their tk paths would be a LOT more comprehensible if they'd use their assigned names.

![image](https://github.com/alejandroautalan/pygubu-designer/assets/1014362/fa7e48b4-edd3-427b-a037-e7dfea412dd3)
Taking the screenshot for the example made it even more obvious that I missed a couple layers that I wasn't aware of.

Maybe it'd be even possible to do both: keeping the old naming look up but also add custom names to the `children` dictionaries.
I was already looking into trying to add that, but I'm still stuck trying to find the point where it's generating / updating a parent's `children`.

**Comment by alejandroautalan** [comment]
Hello @BloodyRain2k ,

I took your idea from alejandroautalan/pygubu#287 and moved to BuilderObject class.

**It is working Ok for named widgets in the designer.**

I tried to enable this for all widgets (auto-generated names and named widgets). But it gives me a problem in the designer preview. I don't know yet what is generating this.

For reference the error is this the following:

```
_tkinter.TclError: bad window path name ".mainframe.mainpw.newmain.nbmain.designer_main.designerpw.frame_1.scrollbarhelper_1.!frame.preview_canvas"
```

Full error:

```
python: 3.11.2 on linux
tk: 8.6
pygubu: 0.31
pygubu-designer: 0.36
Traceback (most recent call last):
  File "/home/alejandro/.local/share/fades/a25be9a4-fb9e-44fc-b58d-45e5388ec816/bin/pygubu-designer", line 8, in <module>
    sys.exit(start_pygubu())
             ^^^^^^^^^^^^^^
  File "/home/alejandro/Privado/projects/pygubu-project/pygubu-designer/src/pygubudesigner/main.py", line 960, in start_pygubu
    app = PygubuDesigner()
          ^^^^^^^^^^^^^^^^
  File "/home/alejandro/Privado/projects/pygubu-project/pygubu-designer/src/pygubudesigner/main.py", line 197, in __init__
    self.previewer = PreviewHelper(
                     ^^^^^^^^^^^^^^
  File "/home/alejandro/Privado/projects/pygubu-project/pygubu-designer/src/pygubudesigner/preview/helper.py", line 58, in __init__
    canvas.bind("<Button-1>", self.click_handler)
  File "/usr/lib/python3.11/tkinter/__init__.py", line 1448, in bind
    return self._bind(('bind', self._w), sequence, func, add)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/tkinter/__init__.py", line 1402, in _bind
    self.tk.call(what + (sequence, cmd))
_tkinter.TclError: bad window path name ".mainframe.mainpw.newmain.nbmain.designer_main.designerpw.frame_1.scrollbarhelper_1.!frame.preview_canvas"

```

Regards
Alejandro A.

**Comment by BloodyRain2k** [comment]
I've done some work on my app that drives most of the changes I come up with and I've ran into the same "bad window path" error, but in an entirely different way:
I've added a 2nd `tk.Toplevel` to my app's `.ui` and noticed, that when such a Toplevel gets closed (aka destroyed by tk) then the `.builder` of the main app will *not* create a new instance when `{appInstance}.builder.get_object("{widget_name}")` is called once more.
Instead it will return the reference to the previously destroyed widget and most tk functions of that destroyed instance then throw the same error.
In my case was that caused by calling `.wm_minsize()` on the widget. I tried out some other functions, to see if any would let me check if the widget was still "alive" without throwing that error, but eventually gave up because it seems that `destroy()` cleaned up everything (as intended as far tk is concerned) but the `.builder` didn't get the update.

Not sure if that error might be related to the change with naming or if it's just coincidence.

But another thing I noticed, was that the widget naming as you implemented it works only for child widgets.
My two `tk.Toplevel` widgets for example, the main window and the dialog I added today, both stay unnamed.
The main window just gets the `._name = "."` and no `.widgetName` property, while the dialog gets `._name = ".!toplevel{n}"` and `.widgetName = "!toplevel{n}"`.
Both are named and do pass those names onto the function that I originally modified to apply the name.

I'm currently looking into both things to see what it's about.
A quick but dirty "fix" for the Builder not instancing a new widget, was to just pop the respective entry on the widget's `<Destroy>`. Works but it's far from ideal...
